### PR TITLE
Fix the five runtime bugs found while building comlint (#529-#533)

### LINKED ISSUE #529: postfix()/eval() crashes on syntactically unbalanced input (e.g. a fragment of a multi-line statement)
## Repro

```
p=raster(0,0,64,64 :rgb 4,4,
```

Feeding just this one physical line (an intentionally-incomplete fragment of a `raster()` call that continues across several more lines — see `src/comdraw/tests/pixelfunc.comt:18`) to `postfix()` and then `eval()`-ing the result segfaults the interpreter:

```
eval("postfix(p=raster(0,0,64,64 :rgb 4,4,)")   -- unclosed paren, trailing comma
```

More generally: `postfix()` (and by extension `eval(postfix(...))`, since `postfix()` is `post_eval` and works off raw tokens) does not validate that its input is a syntactically complete, balanced expression before processing it. A fragment with an unclosed `(` and/or a dangling trailing comma crashes the parser instead of returning an error value or raising a catchable condition.

## Discovered via

Building `comlint` (#524, shipped in PR #528). Its first version fed each physical source line straight to `postfix()`; multi-line statements (very common in this codebase — see `pixelfunc.comt`'s multi-line `raster()` call) crashed it immediately on real files. The shipped fix is entirely client-side: `comlint` now tracks cumulative paren balance (quote- and escape-aware) across physical lines itself and never calls `postfix()`/`eval()` on a fragment it can detect is incomplete.

That workaround only protects `comlint`'s own usage. The underlying crash is still reachable directly and by any other caller that builds `postfix()`/`eval()` input from partial or attacker/user-influenced text.

## Severity

Crash (SIGSEGV) via malformed input to two ordinary, documented commands — no exotic setup needed, just an unbalanced/incomplete expression string, which is easy to construct accidentally (e.g. programmatically slicing a script into lines, as `comlint` originally did).

## Fix direction

`postfix()`'s tokenizer/parser should detect an unbalanced or incomplete expression (unclosed paren, dangling trailing comma/operator, etc.) and fail gracefully — return `nil` or a recognizable error value, matching this codebase's existing convention (e.g. `peek()`/`poke()`'s bounds-check-then-return-`nil` pattern from #520/#525) — rather than crash. Worth checking whether `eval()`'s own direct parsing path (independent of `postfix()`) has the same gap, since scripts also build source text for `eval()` outside of `postfix()`'s output.

## Related

- #524 — comlint's design, where this was found
- PR #528 — comlint v1, which works around this client-side but doesn't fix the root cause
- #520/#525 — the bounds-check/nil-return precedent this fix direction follows

### PR DESCRIPTION
**Pull request by vectaport**
## Summary

Fixes the five ComTerp/comdraw/drawserv runtime bugs uncovered while building `comlint` (#524, shipped in #528) — tracked as #529 through #533. Each was reproduced against the real binary before fixing, and fixed with the narrowest change that closes the gap.

- **#529** — `eval("postfix(...)")` on a syntactically unbalanced/incomplete fragment (e.g. one physical line of a multi-line statement, exactly what `comlint` builds) could segfault by actually misfiring the wrapped expression instead of erroring. `ComTerpServ::run(const char*)` now refuses unbalanced input via a quote/char-literal/escape-aware `()`/`[]`/`{}`-balance pre-check, returning `nil` instead of reaching the parser. Reproduced live under `comdraw` with gdb before the fix (confirmed a real SIGSEGV in `CreateRasterFunc::create_from_rgb`), confirmed fixed after.
- **#530** — `comdraw`/`drawserv -runfile` had no way to forward its own trailing arguments into the script (`arg(n)`/`narg()`), unlike comterp's own `run <file> [args...]` subcommand — `comdraw` even hard-rejected any trailing argv as a usage error. Now forwarded via `set_args()`, mirroring `comterp_/main.c`'s convention (`arg(0)` = script path, `arg(1)..` = trailing args). Documented in both `--help` and `comdraw`'s man page, including the one real constraint (a trailing arg spelled exactly like one of comdraw's own options collides with that option — order otherwise doesn't matter, confirmed directly).
- **#531** — `out_form` (backing every `print()` `%`-conversion) silently truncated any formatted value past 1023 bytes, no error or warning. Fixed to measure via `snprintf`'s own return value and grow only when needed; then improved further to a `static`, only-grows-and-is-reused buffer per call site, so repeated large prints don't pay a fresh allocation each time.
- **#532** — `close()` on a pipe-opened `FileObj` called `fclose()` directly instead of the existing `_pipe`-aware `FileObj::close()`, which is undefined behavior on a `popen()`'d `FILE*` per POSIX. Now routes through `FileObj::close()` (also resetting `_fptr` to prevent a double-close from the destructor) and pushes the real subprocess exit status, matching `shell()`'s existing convention.
- **#533** — `open()` had no atomic create-only-if-absent mode. Initially added an `:excl` keyword, then discovered (and removed, per the issue's own follow-up comment) that it was unnecessary: `modestr` was always passed straight through to `fopen()`/`popen()` unmodified, so `open(path "wx")` already gets `O_CREAT|O_EXCL` via glibc's mode-string extension. Just documented the pass-through instead of adding new keyword surface area.

## Verification

- Reproduced #529's crash live (gdb backtrace) before fixing; confirmed no crash after, `result=nil`.
- Reproduced #530/#531/#532/#533 with direct `comterp`/`comdraw` invocations before and after each fix.
- Full `comterp_/tests/run_all.comt` suite: clean (including a byte-for-byte 0-255 round-trip test that caught a real edge case in the #529 balance-checker's escape handling — `\cX` control-char escapes — before it shipped).
- Full `comdraw/tests/run_all.comt` suite (under Xvfb): clean, including `pixelfunc.comt`, the exact multi-line-`raster()` test file #529's issue referenced.

## Related

Closes #529, #530, #531, #532, #533. Split out from #524/#528's design work; #534-#537 (comlint feature/design issues, not bugs) are left for later per the repo owner's direction.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_019Gpnqd8RDaZQ1LzRVWFYei

---
_Generated by [Claude Code](https://claude.ai/code/session_019Gpnqd8RDaZQ1LzRVWFYei)_

### CODE REVIEW

File: src/ComTerp/comterpserv.c [hunk 1]
```diff
@@ -515,6 +515,62 @@ int ComTerpServ::runfile(const char* filename, boolean popen_flag) {
...
+         raw literal byte that's never itself re-interpreted, even when X
+         happens to be '\', '[' or ']' (control byte 28's own spelling,
+         '\c\', is exactly that case, and broke this same round-trip test
+         the same way before this was added).  Every other backslash
+         escape (\n, \xNN, octal, \\, \', \") only ever escapes a single
+         following character that's never itself paren/bracket/brace/
+         quote-like, so a plain one-byte skip covers those. */
+      if (p[1] == 'c' && p[2] != '\0') { p += 3; continue; }
+      p += 2; continue;
+    }
```
[the remark this answers is not in the document]
Comment (vectaport) [reply src/ComTerp/comterpserv.c]: Confirmed real, fixed in 5a80b16e. The check unconditionally treated any backslash as escaping the next 1-3 characters regardless of quote state, so a bare backslash sitting directly before a real delimiter (e.g. `\(`) got swallowed along with it instead of counted, which could misjudge a genuinely balanced expression as unbalanced.

Backslash escapes are only meaningful inside a string/char literal per `_lexscan.c`'s own state machine (`_backslash_ids`, the dormant backslash-identifier feature, is unconditionally 0 with no setter anywhere in the tree, so today a bare unquoted `\` is just an ordinary character to the real lexer). Gated the escape-handling behind `instr || inchar` so it now only fires inside a literal, matching the real lexer. Verified with a standalone harness isolating the exact `\(3+4)` shape against the fix.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/ComTerp/comterpserv.c [hunk 2]
```diff
@@ -515,6 +515,62 @@ int ComTerpServ::runfile(const char* filename, boolean popen_flag) {
...
+   permanently unclosed by a stray inner ")" has been observed to make
+   this path silently lose track of that outer call and fall through to
+   evaluating -- actually firing -- whatever incomplete expression was
+   left inside it, rather than erroring or leaving it unfired the way
+   postfix()'s own post-eval, never-executes-its-argument contract
+   promises (#529).  Rather than chase that state-machine gap through
+   the shared token-stream parser (used by every other entry point, and
+   already established elsewhere as too deep/risky to safely touch --
+   see #524/#265's writeup), refuse unbalanced input here, at the one
+   entry point where the gap was found, before any of it is parsed. */
```
[the remark this answers is not in the document]
Comment (vectaport) [reply src/ComTerp/comterpserv.c]: This thread is anchored to an outdated diff -- the issue-number references (#529 etc.) were already dropped in bd9b5977, before this review ran. The leftover fix-narration you're pointing at ("confirmed the hard way", "broke this same round-trip test", "before this was added") was still there though; rewritten to state the mechanism plainly in 5a80b16e (same commit as the backslash-handling fix above). `check_comment_rules.py` now reports 0 issue-number violations and 0 narration warnings against current HEAD.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/ComTerp/comterp.c [hunk 3]
```diff
@@ -1789,6 +1789,19 @@ int ComTerp::run(boolean one_expr, boolean nested) {
...
+     outlive a single run() the way _stack does, so a caller invoking run()
+     once per already-delimited chunk of input (ComterpHandler::handle_input(),
+     one call per socket line) leaves whatever the PRIOR call's postfix
+     buffer last held sitting here at entry.  eof() below reads that buffer's
+     last token before this call has parsed anything of its own; a plain
+     trailing TOK_EOF there is this same probe's own doing on every clean
+     exit (case TOK_EOF in _parser.c always emits one), never a sign that
+     THIS call's input is already exhausted, so strip it rather than let
+     eof() mistake it for that and skip this call's loop -- and this call's
+     expression -- entirely. */
```
[the remark this answers is not in the document]
Comment (vectaport) [reply src/ComTerp/comterp.c]: Reworded all three to state the invariant directly instead of describing what would go wrong without the check — will land in the next push.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (vectaport) [reply src/ComTerp/comterp.c]: Tightened further — dropped the caller-specific narrative (ComterpHandler::handle_input(), sockets) entirely and stated it purely as ComTerp::run()'s own contract with eof(): the postfix buffer is instance state that outlives one call, so a leftover trailing TOK_EOF from an earlier call must be cleared before eof() can be trusted for this one. Will land in the next push.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/ComTerp/iofunc.h [hunk 4]
```diff
@@ -67,8 +67,9 @@ class OpenFileFunc : public ComFunc {
     OpenFileFunc(ComTerp*);
 
     virtual void execute();
-    virtual const char* docstring() { 
-      return "fileobj|pipeobj=open([filename [modestr]] :pipe :in :out :err) -- open file command"; }
+    virtual const char* docstring() {
+      return "fileobj|pipeobj=open([filename [modestr]] :pipe :in :out :err) -- "
+        "open file command; modestr r w a r+ w+ a+, plus b/x suffixes"; }
```
[the remark this answers is not in the document]
Comment (vectaport) [reply src/ComTerp/iofunc.h]: Good catch, the second pass dropped that qualifier. Fixed: "modestr r w a r+ w+ a+, plus b suffix or x (fail if exists, w/w+ only)" — will land in the next push.

---
_Generated by [Claude Code](https://claude.ai/code)_
